### PR TITLE
ci: improve caching

### PR DESCRIPTION
**Pull request by shumkov**
## Issue being fixed or feature implemented

We not efficiently using different types of caching. Docker layer cache doesn't work.

## What was done?

- Use artifacts for build js
- Store docker cache mounts in GHA cache
- Store librocksdb in GHA cache
- Store local network data in GHA cache
- Store yarn unplugged in GHA cache
- Remove unnecessary "Prebuild and cache some Rust dependencies" job
- Improve sccache vars in Dockerfile

## How Has This Been Tested?

Running CI

## Breaking Changes

None

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [x] I have assigned this pull request to a milestone

## Summary by CodeRabbit

- **New Features**
	- Improved caching mechanisms for various workflows, enhancing performance and reliability.
	- Introduced a new step to configure `sccache` for better caching management in Rust jobs.
	- Added a new job for artifact removal to streamline workflow execution.
	- Added a new input parameter for restoring local network data in the Dashmate tests workflow.
	- Enhanced handling of JavaScript build artifacts through updated workflows.
	- Introduced a new constant to track the last block time in the metrics module.
	- Updated workflows to ensure JavaScript packages are built before running end-to-end tests.
	- Updated workflows to utilize new actions for artifact downloading and checking.
	- Added a new input parameter for caching in the image build workflow.
	- Enhanced caching strategy for Docker images with new AWS integration.
	- Added new public modules in the Rust library for metrics and testing.

- **Bug Fixes**
	- Ensured workflows fail gracefully if caches are not found by adding `fail-on-cache-miss` parameters.

- **Documentation**
	- Updated action descriptions and parameters for clarity and improved usability.

- **Refactor**
	- Streamlined caching processes by replacing outdated actions with the latest `actions/cache@v4`.
	- Simplified cache restoration configurations by removing unnecessary parameters related to S3 buckets.
	- Updated job dependencies to ensure proper execution order.
	- Enhanced artifact management by transitioning to new actions for downloading and checking artifacts.
	- Adjusted workflow configurations to remove outdated caching actions and improve artifact handling.
	- Updated Dockerfile to refine caching mechanisms and enhance build configurations.

- **Chores**
	- Removed unnecessary parameters related to S3 buckets from caching configurations.
	- Updated workflows to utilize the latest actions for downloading and managing artifacts.
	- Deleted the obsolete workflow for rebuilding cached dependencies.

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -236,16 +245,11 @@ jobs:
...
         with:
-          bucket: multi-runner-cache-x1xibo9c
-          root: actions-cache
-          path: build-js-artifacts-${{ github.sha }}.tar
-          key: build-js-artifacts/${{ github.sha }}
-
-      - name: Unpack JS build artifacts archive
-        run: tar -xf build-js-artifacts-${{ github.sha }}.tar
+          name: js-build-${{ github.sha }}
+          path: packages
```
[the remark this answers is not in the document]
Comment (shumkov) [reply .github/workflows/release.yml:244]: there is no such option

File: .github/workflows/tests-build-js.yml [hunk 2]
```diff
@@ -13,55 +13,59 @@ jobs:
...
-        uses: strophy/actions-cache@opendal-update
-        id: cache
+      - uses: softwareforgood/check-artifact-v4-existence@v0
+        id: check-artifact
         with:
-          bucket: multi-runner-cache-x1xibo9c
-          root: actions-cache
-          path: build-js-artifacts-${{ github.sha }}.tar
-          key: build-js-artifacts/${{ github.sha }}
+          name: js-build-${{ github.sha }}
```
[the remark this answers is not in the document]
Comment (shumkov) [reply .github/workflows/tests-build-js.yml:19]: There is not such option

File: .github/actions/s3-layer-cache-settings/action.yaml [hunk 3]
```diff
@@ -46,21 +46,24 @@ runs:
           const manifestNames = [
             '${{ inputs.name }}_sha_${{ github.sha }}',
             `${{ inputs.name }}_tag_${ sanitizedHeadRef }`,
-            '${{ inputs.name }}'
+            '${{ inputs.name }}' // TODO: must be only on default branch
```
Comment (lklimek) [inline .github/actions/s3-layer-cache-settings/action.yaml:49]: TODO

File: .github/workflows/tests.yml [hunk 4]
```diff
@@ -97,13 +77,6 @@ jobs:
...
     with:
       package: ${{ matrix.rs-package }}
-      # lint-runner: ${{ contains(fromJSON('["drive-abci", "drive"]'), matrix.rs-package) && '[ "self-hosted", "linux", "arm64", "ubuntu-platform" ]' || '"ubuntu-24.04"' }}
-      # FIXME: Clippy fails on github hosted runners, most likely due to RAM usage. Using self-hosted runners for now.
-      # lint-runner: '[ "self-hosted", "linux", "arm64", "ubuntu-platform" ]'
-      lint-runner: '["ubuntu-24.04"]'
-      # Run drive tests on self-hosted 4x
-      # test-runner: '[ "self-hosted", "linux", "arm64", "ubuntu-platform" ]'
-      test-runner: '["ubuntu-24.04"]'
       check-each-feature: ${{ contains(fromJSON('["dash-sdk","rs-dapi-client","dapi-grpc","dpp","drive-abci"]'), matrix.rs-package) }}
```
Comment (lklimek) [inline .github/workflows/tests.yml:80]: ubuntu-24.04 ?

Comment (shumkov) [reply .github/workflows/tests.yml:80]: Yes, it uses ubuntu 24.04

File: .github/actions/docker/action.yaml [hunk 5]
```diff
@@ -90,16 +77,82 @@ runs:
...
+      uses: actions/cache@v4
+      id: cargo-cache
+      with:
+        path: |
+          ${{ env.HOME }}/cargo-cache-registry-index
+          ${{ env.HOME }}/cargo-cache-registry-cache
+          ${{ env.HOME }}/cargo-cache-git-db
+        key: ${{ runner.os }}-cargo-${{ hashFiles('Cargo.lock') }}
+        restore-keys: |
+          ${{ runner.os }}-cargo-
```
[the remark this answers is not in the document]
Comment (shumkov) [reply .github/actions/docker/action.yaml:97]: It uses `key` first and then `restore-keys` prefixes

**Review by lklimek — APPROVED** [review]

File: .github/actions/docker/action.yaml [hunk 6]
```diff
@@ -37,6 +37,9 @@ inputs:
   aws_secret_access_key:
     description: AWS secret access key
     required: true
+  cache_to_name:
```
[the remark this answers is not in the document]
Comment (shumkov) [reply .github/actions/docker/action.yaml:40]: We are autodetecting it in workflow. It's not a good idea to autodetect and have some magic inside actions.

File: .github/workflows/tests-build-image.yml [hunk 7]
```diff
@@ -48,3 +48,4 @@ jobs:
           region: ${{ secrets.AWS_REGION }}
           aws_access_key_id: ${{ secrets.AWS_ACCESS_KEY_ID }}
           aws_secret_access_key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          cache_to_name: ${{ github.event_name == 'push' && 'true' || 'false' }}
```
[the remark this answers is not in the document]
Comment (shumkov) [reply .github/workflows/tests-build-image.yml:51]: I fixed that. Thank you

**Review by lklimek — APPROVED** [review]

**Pull request merged**